File: common-core/src/main/java/io/quartic/common/client/ClientBuilder.java
package io.quartic.common.client;

import feign.Feign;
import feign.Logger;
import feign.Retryer;
import feign.jackson.JacksonDecoder;
import feign.jackson.JacksonEncoder;
import feign.jaxrs.JAXRSContract;
import feign.slf4j.Slf4jLogger;

import static com.google.common.net.HttpHeaders.USER_AGENT;
import static io.quartic.common.serdes.ObjectMappersKt.objectMapper;

public final class ClientBuilder {
    private ClientBuilder() {}

    public static <T> T build(Class<T> target, Class<?> owner, String url) {
        return build(target, Utils.userAgentFor(owner), url);
    }

    private static <T> T build(Class<T> target, String userAgent, String url) {
        return Feign.builder()
                .contract(new JAXRSContract())
                .encoder(new InputStreamEncoder(new JacksonEncoder(objectMapper())))
                .decoder(new JacksonDecoder(objectMapper()))
                .retryer(new Retryer.Default(0, 0, 1))
                .requestInterceptor(template -> template.header(USER_AGENT, userAgent))
                .logger(new Slf4jLogger(ClientBuilder.class))
                .logLevel(Logger.Level.BASIC)
                .target(target, url);
    }

}

File: common-core/src/main/kotlin/io/quartic/common/client/ClientBuilder.kt
package io.quartic.common.client

import io.quartic.common.ApplicationDetails
import io.quartic.common.serdes.OBJECT_MAPPER
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.java8.Java8CallAdapterFactory
import retrofit2.converter.jackson.JacksonConverterFactory
import java.net.URI
import java.util.concurrent.TimeUnit


class ClientBuilder(val owner: Class<*>) {
    constructor(owner: Any) : this(owner.javaClass)

    inline fun <reified T> retrofit(url: URI, timeoutSeconds: Long = RETROFIT_DEFAULT_TIMEOUT): T =
        retrofit(url.toString(), timeoutSeconds)
    inline fun <reified T> retrofit(url: String, timeoutSeconds: Long = RETROFIT_DEFAULT_TIMEOUT): T {
        val target = T::class.java
        if (!target.isAnnotationPresent(Retrofittable::class.java)) {
            throw IllegalArgumentException("${target.simpleName} is not marked as @${Retrofittable::class.simpleName}")
        }

        val urlWithSlash = if(url.endsWith("/")) url else url + "/"
        val interceptor = HttpLoggingInterceptor()
        interceptor.level = HttpLoggingInterceptor.Level.BASIC
        val client = OkHttpClient.Builder()
            .readTimeout(timeoutSeconds, TimeUnit.SECONDS)
            .addInterceptor(interceptor)
            .addInterceptor { chain ->
                val original = chain.request()

                val request = original.newBuilder()
                    .header("User-Agent", userAgentFor(owner))
                    .method(original.method(), original.body())
                    .build()

                chain.proceed(request)
            }
            .build()

        return Retrofit.Builder()
            .baseUrl(urlWithSlash)
            .client(client)
            .addConverterFactory(JacksonConverterFactory.create(OBJECT_MAPPER))
            .addCallAdapterFactory(Java8CallAdapterFactory.create())
            .build()
            .create(target)
    }

    companion object {
        @Target(AnnotationTarget.CLASS)
        annotation class Retrofittable
        @Target(AnnotationTarget.CLASS)
        annotation class Jaxable

        val RETROFIT_DEFAULT_TIMEOUT: Long = 10


        fun userAgentFor(clazz: Class<*>): String {
            val details = ApplicationDetails(clazz)
            return "Quartic-${details.name}/${details.version}"
        }
    }
}
